File: upande_accounting/upande_accounting_customizations/report/creditors_aging/creditors_aging.js
// Copyright (c) 2026, jeniffer@upande.com and contributors
// For license information, please see license.txt

frappe.query_reports["Creditors Aging"] = {

    // ------------------------------------------------------------------
    // On load
    // ------------------------------------------------------------------
    onload: function (report) {
        report.page.add_inner_button(__("Creditors Aging Summary"), function () {
            frappe.route_options = {
                company:           frappe.query_report.get_filter_value("company"),
                report_date:       frappe.query_report.get_filter_value("report_date"),
                ageing_based_on:   frappe.query_report.get_filter_value("ageing_based_on"),
                range:             frappe.query_report.get_filter_value("range"),
                party:             frappe.query_report.get_filter_value("party"),
                in_party_currency: frappe.query_report.get_filter_value("in_party_currency") || 0,
                include_draft:     frappe.query_report.get_filter_value("include_draft") || 0,
            };
            frappe.set_route("query-report", "Creditors Aging Summary");
        });
    },

    // ------------------------------------------------------------------
    // Filters
    // ------------------------------------------------------------------
    filters: [
        {
            fieldname: "company",
            label:     __("Company"),
            fieldtype: "Link",
            options:   "Company",
            reqd:      1,
            default:   frappe.defaults.get_user_default("Company"),
        },
        {
            fieldname: "report_date",
            label:     __("As On Date"),
            fieldtype: "Date",
            reqd:      1,
            default:   frappe.datetime.get_today(),
        },
        {
            fieldname: "ageing_based_on",
            label:     __("Ageing Based On"),
            fieldtype: "Select",
            options:   "Posting Date\nDue Date",
            default:   "Due Date",
        },
        {
            fieldname: "range",
            label:     __("Ageing Range (days)"),
            fieldtype: "Data",
            default:   "30, 60, 90, 120",
        },
        {
            fieldname: "party_account",
            label:     __("Payable Account"),
            fieldtype: "Link",
            options:   "Account",
            get_query: () => {
                const company = frappe.query_report.get_filter_value("company");
                return { filters: { company, account_type: "Payable", is_group: 0 } };
            },
        },
        {
            fieldname: "party",
            label:     __("Supplier(s)"),
            fieldtype: "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options("Supplier", txt);
            },
        },
        {
            fieldname: "supplier_group",
            label:     __("Supplier Group"),
            fieldtype: "Link",
            options:   "Supplier Group",
        },
        {
            fieldname: "payment_terms_template",
            label:     __("Payment Terms Template"),
            fieldtype: "Link",
            options:   "Payment Terms Template",
        },
        {
            fieldname: "finance_book",
            label:     __("Finance Book"),
            fieldtype: "Link",
            options:   "Finance Book",
        },
        {
            fieldname: "cost_center",
            label:     __("Cost Center"),
            fieldtype: "Link",
            options:   "Cost Center",
            get_query: () => {
                const company = frappe.query_report.get_filter_value("company");
                return { filters: { company } };
            },
        },
        {
            fieldname: "based_on_payment_terms",
            label:     __("Based On Payment Terms"),
            fieldtype: "Check",
        },
        {
            fieldname: "show_future_payments",
            label:     __("Show Future Payments"),
            fieldtype: "Check",
        },
        {
            fieldname: "show_remarks",
            label:     __("Show Remarks"),
            fieldtype: "Check",
        },
        {
            fieldname: "in_party_currency",
            label:     __("In Party Currency"),
            fieldtype: "Check",
            default:   0,
        },
        {
            fieldname: "include_draft",
            label:     __("Include Draft Invoices"),
            fieldtype: "Check",
            default:   0,
        },
    ],

    // ------------------------------------------------------------------
    // Formatter
    // ------------------------------------------------------------------
    formatter: function (value, row, column, data, default_formatter) {
        if (!data) return default_formatter(value, row, column, data);

        value = default_formatter(value, row, column, data);

        if (data.bold) {
            value = `<strong>${value}</strong>`;
        }

        // Status column — colour-coded by urgency
        if (column.fieldname === "status" && data.status && !data.is_draft) {
            const palette = {
                "Overdue":    { color: "#c0392b", bg: "#fdecea" },
                "Almost Due": { color: "#d35400", bg: "#fef3e2" },
                "Not Due":    { color: "#1e8449", bg: "#eafaf1" },
            };
            const style = palette[data.status];
            if (style) {
                value = `<span style="
                    color:${style.color};
                    background:${style.bg};
                    border-radius:3px;
                    padding:1px 6px;
                    font-weight:500;
                    font-size:0.85em;">${value}</span>`;
            }
        }

        // Draft invoice rows — orange italic across all cells
        if (data.is_draft) {
            value = `<span style="color:#e67e22; font-style:italic;">${value || ""}</span>`;
        }

        return value;
    },

    get_datatable_options(options) {
        return Object.assign(options, { checkboxColumn: true });
    },
};

File: upande_accounting/upande_accounting_customizations/report/creditors_aging/creditors_aging.py
# Copyright (c) 2026, jeniffer@upande.com and contributors
# For license information, please see license.txt

"""
Creditors Aging — per-invoice aging report (mirrors Accounts Payable).

Inherits the full ReceivablePayableReport logic from ERPNext and adds:

  1. Custom column layout  — Bill No and Bill Date as primary identifiers;
     Voucher No moved to the last column.

  2. Status column  — Overdue / Almost Due / Not Due computed from due_date
     vs the As On Date (report_date).

  3. include_draft  — appends unsubmitted Purchase Invoices as individual rows
     (orange italic), each placed in the correct aging bucket.

  4. in_party_currency  — exposed as an explicit filter; handled by the base.

Navigation: "Creditors Aging Summary" button switches to the per-supplier
summary carrying all filters across.
"""

import frappe
from frappe import _, scrub
from frappe.utils import flt, getdate

from erpnext.accounts.report.accounts_receivable.accounts_receivable import ReceivablePayableReport


def execute(filters=None):
    filters = frappe._dict(filters or {})
    filters.party_type = "Supplier"
    args = {
        "account_type": "Payable",
        "naming_by": ["Buying Settings", "supp_master_name"],
    }
    return CreditorsAgingReport(filters).run(args)


class CreditorsAgingReport(ReceivablePayableReport):
    # ------------------------------------------------------------------
    # Columns
    # ------------------------------------------------------------------

    def add_column(self, label, fieldname=None, fieldtype="Currency", options=None, width=120):
        if not fieldname:
            fieldname = scrub(label)
        if fieldtype == "Currency":
            options = "currency"
        self.columns.append(
            dict(label=label, fieldname=fieldname, fieldtype=fieldtype, options=options, width=width)
        )

    def get_columns(self):
        self.columns = []
        self.add_column(_("Posting Date"), fieldname="posting_date", fieldtype="Date",        width=120)
        self.add_column(_("Party Type"),   fieldname="party_type",   fieldtype="Data",        width=100)
        self.add_column(_("Party"),        fieldname="party",        fieldtype="Dynamic Link",
                        options="party_type", width=250)
        self.add_column(_("Voucher Type"), fieldname="voucher_type", fieldtype="Data",        width=130)
        self.add_column(_("Bill No"),      fieldname="bill_no",      fieldtype="Data",        width=150)
        self.add_column(_("Bill Date"),    fieldname="bill_date",    fieldtype="Date",        width=120)
        self.add_column(_("Due Date"),     fieldname="due_date",     fieldtype="Date",        width=120)
        self.add_column(_("Status"),       fieldname="status",       fieldtype="Data",        width=110)
        self.add_column(_("Currency"),     fieldname="currency",     fieldtype="Link",
                        options="Currency", width=80)

        if self.filters.based_on_payment_terms:
            self.add_column(_("Payment Term"),        fieldname="payment_term",        fieldtype="Data", width=150)
            self.add_column(_("Invoice Grand Total"), fieldname="invoice_grand_total",                   width=150)

        self.add_column(_("Paid Amount"),        fieldname="paid",        width=130)
        self.add_column(_("Invoiced Amount"),    fieldname="invoiced",    width=130)
        self.add_column(_("Debit Note"),         fieldname="credit_note", width=130)
        self.add_column(_("Outstanding Amount"), fieldname="outstanding", width=150)
        self.add_column(_("Age (Days)"),         fieldname="age",         fieldtype="Int", width=80)
        self.setup_ageing_columns()

        self.add_column(_("Supplier Group"), fieldname="supplier_group", fieldtype="Link",
                        options="Supplier Group", width=130)

        if self.filters.show_remarks:
            self.add_column(_("Remarks"), fieldname="remarks", fieldtype="Text", width=200)

        if self.filters.show_future_payments:
            self.add_column(_("Future Payment Ref"),    fieldname="future_ref",        fieldtype="Data", width=150)
            self.add_column(_("Future Payment Amount"), fieldname="future_amount",                       width=150)
            self.add_column(_("Remaining Balance"),     fieldname="remaining_balance",                   width=150)

        # Voucher No last
        self.add_column(_("Voucher No"), fieldname="voucher_no", fieldtype="Dynamic Link",
                        options="voucher_type", width=180)

    # ------------------------------------------------------------------
    # Status helpers
    # ------------------------------------------------------------------

    def _set_row_status(self, row):
        due_date = row.get("due_date")
        if not due_date:
            row.status = ""
            return
        days_until_due = (getdate(due_date) - getdate(self.filters.report_date)).days
        if days_until_due > 7:
            row.status = "Not Due"
        elif days_until_due > 0:
            row.status = "Almost Due"
        else:
            row.status = "Overdue"

    def append_row(self, row):
        super().append_row(row)
        self._set_row_status(row)

    # ------------------------------------------------------------------
    # Data
    # ------------------------------------------------------------------

    def get_data(self):
        super().get_data()
        if self.filters.get("include_draft"):
            self._append_draft_rows()

    def _append_draft_rows(self):
        company      = self.filters.company
        report_date  = getdate(self.filters.report_date)
        in_party_cur = self.filters.get("in_party_currency")

        sql_filters = {
            "docstatus":    0,
            "company":      company,
            "posting_date": ("<=", report_date),
        }
        if self.filters.get("party"):
            sql_filters["supplier"] = ("in", self.filters.get("party"))

        draft_pis = frappe.get_all(
            "Purchase Invoice",
            filters=sql_filters,
            fields=[
                "name", "supplier", "posting_date", "due_date",
                "bill_no", "bill_date",
                "currency", "grand_total", "base_grand_total",
            ],
        )

        for pi in draft_pis:
            amount = flt(pi.grand_total) if in_party_cur else flt(pi.base_grand_total)
            supp_currency = (
                frappe.db.get_value("Supplier", pi.supplier, "default_currency")
                or self.company_currency
            )

            row = frappe._dict(
                posting_date        = pi.posting_date,
                party_type          = "Supplier",
                party               = pi.supplier,
                party_account       = "",
                voucher_type        = "Purchase Invoice",
                voucher_no          = pi.name,
                due_date            = pi.due_date,
                bill_no             = pi.bill_no or "",
                bill_date           = pi.bill_date,
                invoiced            = amount,
                invoice_grand_total = amount,
                paid                = 0.0,
                credit_note         = 0.0,
                outstanding         = amount,
                account_currency    = supp_currency,
                currency            = supp_currency if in_party_cur else self.company_currency,
                is_draft            = True,
            )

            party_details = self.get_party_details(pi.supplier) or {}
            row.update(party_details)
            self.set_ageing(row)
            self._set_row_status(row)

            self.data.append(row)
